### PR TITLE
Player: `Player_ActionHandler_Talk`

### PR DESCRIPTION
**Pull request by fig02**
Part 1 of a few related to talking.
There are some actor flags and state flags I have named in a separate branch but Ill do those in following PRs.

I'm usually not a fan of goto's but it heavily improves the structure of this function. Recommend reading through it on its own, not in githubs diff view.

### CODE REVIEW

File: include/z64player.h [hunk 1]
```diff
@@ -697,7 +697,7 @@ typedef struct WeaponInfo {
 #define PLAYER_STATE1_31 (1 << 31)
 
 #define PLAYER_STATE2_0 (1 << 0)
-#define PLAYER_STATE2_1 (1 << 1)
+#define PLAYER_STATE2_CAN_ACCEPT_TALK_OFFER (1 << 1) // Can talk to an actor. "Speak" or "Check" is shown on the A button.
```
Comment (cadmic) [inline include/z64player.h:700]: `PLAYER_STATE2_CAN_ACCEPT_TALK_OFFER`? or is there a time where player can talk without accepting a talk offer or vice versa?

Comment (fig02) [reply include/z64player.h:700]: Your suggestion is the same as what I have, did you forget to type something else

Comment (cadmic) [reply include/z64player.h:700]: yes, sorry. `PLAYER_STATE2_CAN_TALK`?

Comment (fig02) [reply include/z64player.h:700]: Yeah any time talking is initiated through c-up it is not done through the talk offer mechanism. I can update the comment to be more specific

**Review by cadmic — APPROVED** [review]

**Review by krm01 — APPROVED** [review]
it's sad that `goto dont_talk` for all of the exits doesn't match, or `return false` for all of them similarly. the fact that both exist is evidence that the author didn't actually use the gotos, but I agree that its easier to read this way.

**Comment by fig02** [comment]
yeah do agree there. player dev loved his complex conditions so I dont doubt it was a jumbled mess originally. but yeah since it matches all versions like this, i think we take the readability improvement

File: src/overlays/actors/ovl_player_actor/z_player.c [hunk 2]
```diff
@@ -5941,76 +5941,107 @@ s32 Player_ActionHandler_13(Player* this, PlayState* play) {
...
+            talkOfferActor = cUpTalkActor;
+            this->talkActor = NULL;
+
+            if (forceTalkToNavi || !canTalkToLockOnWithCUp) {
+                cUpTalkActor->textId = ABS(this->naviTextId);
+            } else {
+                if (cUpTalkActor->naviEnemyId != NAVI_ENEMY_NONE) {
+                    cUpTalkActor->textId = cUpTalkActor->naviEnemyId + 0x600;
                 }
             }
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_player_actor/z_player.c:6026]: Very nitpicky, merge the else if for symmetry on the assignments
```suggestion
            if (forceTalkToNavi || !canTalkToLockOnWithCUp) {
                cUpTalkActor->textId = ABS(this->naviTextId);
            } else if (cUpTalkActor->naviEnemyId != NAVI_ENEMY_NONE) {
                cUpTalkActor->textId = cUpTalkActor->naviEnemyId + 0x600;
            }
```

**Review by AngheloAlf — APPROVED** [review]
The gotos really improve readability.
Just a very minor nit, but I'll approve anyways

File: src/overlays/actors/ovl_player_actor/z_player.c [hunk 3]
```diff
@@ -5941,76 +5941,107 @@ s32 Player_ActionHandler_13(Player* this, PlayState* play) {
...
-                        }
+        if (talkOfferActor != NULL) {
+            // At this point the talk offer can be accepted.
+            // "Speak" or "Check" will appear on the A button in the HUD.
+            this->stateFlags2 |= PLAYER_STATE2_CAN_ACCEPT_TALK_OFFER;
 
-                        sp34 = sp2C;
-                        this->talkActor = NULL;
+            if (CHECK_BTN_ALL(sControlInput->press.button, BTN_A) || (talkOfferActor->flags & ACTOR_FLAG_16)) {
+                // Clearing `cUpTalkActor` gurantees that `talkOfferActor` is the actor that will be spoken to
```
Comment (engineer124) [inline src/overlays/actors/ovl_player_actor/z_player.c:6001]:
```suggestion
                // Clearing `cUpTalkActor` guarantees that `talkOfferActor` is the actor that will be spoken to
```

File: src/overlays/actors/ovl_player_actor/z_player.c [hunk 4]
```diff
@@ -5941,76 +5941,107 @@ s32 Player_ActionHandler_13(Player* this, PlayState* play) {
...
+    if (canTalkToLockOnWithCUp || (this->naviTextId != 0)) {
+        // If `naviTextId` is negative and outside the 0x2XX range, talk to navi instantly
+        forceTalkToNavi = (this->naviTextId < 0) && ((ABS(this->naviTextId) & 0xFF00) != 0x200);
+
+        if (forceTalkToNavi || !canTalkToLockOnWithCUp) {
+            // If `lockOnActor` can't be talked to with c-up, the only option left is navi
+            cUpTalkActor = this->naviActor;
+
+            if (forceTalkToNavi) {
+                // Clearing these pointers gurantees that `cUpTalkActor` will take priority
```
Comment (engineer124) [inline src/overlays/actors/ovl_player_actor/z_player.c:5964]:
```suggestion
                // Clearing these pointers guarantees that `cUpTalkActor` will take priority
```

**Review by engineer124 — APPROVED** [review]

File: src/overlays/actors/ovl_player_actor/z_player.c (hunk 4 above)
Comment (fig02) [reply src/overlays/actors/ovl_player_actor/z_player.c:5964]: damn i cannot spell that word apparently

**Pull request merged**